server/src/main/java/io/crate/planner/operators/EquiJoinDetector.java

```diff
@@ -86,7 +86,15 @@ public Void visitFunction(Function function, Context context) {
             String functionName = function.name();
             switch (functionName) {
                 case NotPredicate.NAME -> {
-                    return null;
+                    if (context.insideEqualOperand) {
+                        // Not a top-level expression but inside EQ operator.
+                        // We need to collect all relations.
+                        for (Symbol arg : function.arguments()) {
+                            arg.accept(this, context);
+                        }
+                    } else {
+                        return null;
+                    }
```

Comment:
Theoretically, OR also needs special handling.  
Since we short circuit to `NestedLoopJoin` it doesn't hurt correctness, only leads false negative for HashJoin.
See https://github.com/crate/crate/pull/16200#issuecomment-2185639823. 

Basically OR is already aligned with the suggestion to "always go with NestedLoopJoin when EQ is applied for boolean operands.

Reply:
False negatives for HashJoin have serious performance implications. Hash-joins have the complexity of O(n) while Nested-Loops join O(n^2).

Reply:
Maybe we could do a follow-up here ?

Reply:
>False negatives for HashJoin have serious performance implications. Hash-joins have the complexity of O(n) while Nested-Loops join O(n^2).

I know, but for booleans not really. We have only 2 keys in map and anything else is collisions. When joining 2 big tables on boolean expression, we would have 2 giant lists with collisions and basically it will be very similar to NestedLoopJoin due to many collisions with extra memory on top of it. 

Of course this has to be verified first. I will create a card and bring this to the refinement.

Reply:
I will create a card for it

Reply:
I debugged a bit to confirm what exactly we are hashing in such scenarios (join on boolean expressions), we hash not int column values but boolean expression results [here](https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/engine/join/HashJoinOperation.java#L120). 

Because of that, O(N) pretty much becomes O(N^2). We basically have 2 LinkedLists , one filled by "true" and another by "false". 

I played around with benchmarks locally. 

First some simple join on int column comparison, to have a baseline how much better HashJoin could be for a normal use case on specific table size
```
[setup]
statements = [
    "create table t1 (c1 int)",
    "insert into t1 (c1) select b from generate_series(1, 1000) a(b)",
    "create table t2 (c2 int)",
    "insert into t2 (c2) select b from generate_series(1, 1000) a(b)",
    "refresh table t1, t2"
]

[[queries]]
statement = "select c1, c2 from t1 join t2 on t1.c1 = t2.c2"
iterations = 100

[teardown]
statements = ["drop table if exists t1", "drop table if exists t2"]
```

This is result for CrateDB 5.7.3
```
## Running Query:
    Statement:
    select c1, c2 from doc.t1 join doc.t2 on t1.c1 = t2.c2
    Concurrency: 1
    Iterations: 100
        100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100/100 [00:00<00:00, 140.04 requests/s]
    Runtime (in ms):
    mean:    5.093 ± 1.780
    min/max: 2.684 → 93.937
    Percentile:
        50:   3.961 ± 9.083 (stdev)
        95:   7.246
        99.9: 93.937
```

Then I tweaked `RewriteJoinPlan` to always go with `NestedLoopJoin` and used 
` cr8 run-crate ./ -- run-spec ../crate-benchmarks/specs/select/hash_joins.toml {node.http_url}`
and got much worse results (as expected)

```
    ## Running Query:
    Statement:
    select c1, c2 from doc.t1 join doc.t2 on t1.c1 = t2.c2
    Concurrency: 1
    Iterations: 100
        100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100/100 [00:08<00:00, 11.24 requests/s]
    Runtime (in ms):
    mean:    86.674 ± 3.271
    min/max: 81.733 → 247.900
    Percentile:
        50:   83.584 ± 16.686 (stdev)
        95:   92.129
        99.9: 247.900
```

Difference is huge, mean is 5 vs 85.

Now I tried the same for boolean expression join

```
[setup]
statements = [
    "create table t1 (c1 int)",
    "insert into t1 (c1) select b from generate_series(1, 1000) a(b)",
    "create table t2 (c2 int)",
    "insert into t2 (c2) select b from generate_series(1, 1000) a(b)",
    "refresh table t1, t2"
]

[[queries]]
statement = "select c1, c2 from t1 join t2 on (t1.c1 > 500) = (t2.c2 < 500)"
iterations = 100

[teardown]
statements = ["drop table if exists t1", "drop table if exists t2"]
```

CrateDB 5.7.3
```
## Running Query:
    Statement:
    select c1, c2 from t1 join t2 on (t1.c1 > 500) = (t2.c2 < 500)
    Concurrency: 1
    Iterations: 100
        100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100/100 [00:45<00:00,  2.19 requests/s]
    Runtime (in ms):
    mean:    126.268 ± 7.421
    min/max: 116.537 → 474.341
    Percentile:
        50:   119.783 ± 37.864 (stdev)
        95:   146.153
        99.9: 474.341
        # Running tearDown
```
and CrateDB with hash join disabled
` cr8 run-crate ./ -- run-spec ../crate-benchmarks/specs/select/hash_joins.toml {node.http_url}`

```
    ## Running Query:
    Statement:
    select c1, c2 from t1 join t2 on (t1.c1 > 500) = (t2.c2 < 500)
    Concurrency: 1
    Iterations: 100
        100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100/100 [00:53<00:00,  1.88 requests/s]
    Runtime (in ms):
    mean:    195.707 ± 7.555
    min/max: 184.861 → 501.112
    Percentile:
        50:   188.567 ± 38.546 (stdev)
        95:   203.298
        99.9: 501.112
```

Here difference is not that hyge anymore, mean 126 vs 195. HashJoin is slightly better because it cuts at least half of entries, HashMap has 2 keys, ie we still have 2 LinkedLists. But in O-notation it's same.

Reply:
BTW https://github.com/crate/crate/issues/16230 is still valid because:
1. Condition simplification can save os of potential false positives, ie not performance problems but actual bugs
2. As mentioned in https://github.com/crate/crate/issues/15443, 
>  One benefit of it would be higher utilization of Get operation

Reply:
> Theoretically, OR also needs special handling.

But I can also add this extra handling here, to get this 1.5x improvement. wdyt?

Reply:
Interesting results. The hash-join should perform better with larger n. This is a single-node setup, right ? Then it is a nested-block-join loop. I think the numbers will be very different on larger n and in a distributed setup. The hashing could be improved.

Reply:
> I can also add this extra handling here, to get this 1.5x improvement. Not huge as it could be in "normal" use case but still notable. wdyt?

What would the special or-handling do ?

Reply:
>What would the special or-handling do ?

It would be pretty much same change as this PR but applied to  `OrOperator.NAME `(instead of immediately short circuiting to false).

Reply:
This fix prevents a false-positive for NOT in the join-condition. So there is a case for `OR` which can be executed as a hash-join but it's executed as a nested-loop-join ?

Reply:
>So there is a case for OR which can be executed as a hash-join but it's executed as a nested-loop-join ?
yes, any join on boolean expressions with OR

So there is a case for OR which can be executed as a hash-join but it's executed as a nested-loop-join ?
Even smth simple like `... ON (t1.c1 > 500) = (t2.c2 < 500 or t2.c2 %2  = 0)`

Reply:
I see, then either file an issue or open a pr directly.

Reply:
https://github.com/crate/crate/pull/16320